system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,513 @@
+//            {
+//                delete _popCommData(i);
+//            }
+//        }
+//    }
+
+    asyncCommData.push_back(commData);
+}
+
+CommData* popCommData(int index)
+{
+    _TF("popCommData", index);
+    CriticalBlock block(commDataLock);
+    return _popCommData(index);
+}
+
+std::vector<CommData*> popCommData(bool send, int rank, int tag, MPI::Comm& comm)
+{
+   _TF("popCommData", rank, tag);
+   CriticalBlock block(commDataLock);
+   std::vector<CommData*> matchedResults = std::vector<CommData*>();
+   int size = asyncCommData.size();
+   for(int i=(size-1); i>=0; i--)
+   {
+       if (asyncCommData[i]->isEqual(send, rank, tag, comm))
+           matchedResults.push_back(_popCommData(i));
+   }
+   return matchedResults;
+}
+
+CommData* popCommData(CommData * commData)
+{
+    _TF("popCommData(CommData * commData)", commData->rank, commData->tag);
+    CriticalBlock block(commDataLock);
+    int size = asyncCommData.size();
+    for(int i=(size-1); i>=0; i--)
+    {
+        if (asyncCommData[i]==commData)
+        {
+            _popCommData(i);
+            break;
+        }
+    }
+    return commData;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------//
+
+#define TAG_UB 100000 //INT_MAX //MPI_TAG_UB
+int getRank(rank_t sourceRank)
+{
+    _TF("getRank", sourceRank);
+    if (sourceRank == RANK_ALL)
+        return MPI::ANY_SOURCE;
+    else
+        return sourceRank;
+}
+
+int getTag(mptag_t mptag)
+{
+    _TF("getTag", mptag);
+    int returnVal;
+    unsigned tag = mptag;
+    if (mptag == TAG_ALL)
+        returnVal = MPI::ANY_TAG;
+    else
+    {
+        if (tag > TAG_UB)
+        {
+            tag= TAG_UB-((unsigned)-tag);                         //MPI doesn't allow custom tags with negative values or beyond MPI_TAG_UB. Thus we shift it to the MPI_TAG_UB range
+        }
+        returnVal = static_cast<int>(tag);
+
+    }
+    return returnVal;
+}
+
+mptag_t getTag(int tag)
+{
+    _TF("getTag", tag);
+    unsigned returnVal;
+    if (tag == MPI::ANY_TAG)
+        returnVal = TAG_ALL;
+    else
+    {
+        if (tag > (TAG_UB/2))
+        {
+            returnVal = (unsigned) - (TAG_UB-tag);                         //MPI doesn't allow custom tags with negative values
+        } else
+            returnVal = tag;
+
+    }
+    return (mptag_t)returnVal;
+}
+
+MPI::Status waitToComplete( bool& completed, bool& error, bool& canceled, bool& timedout, unsigned timeout, CommData *commData)
+{
+    _TF("waitToComplete", completed, error, canceled, timeout);
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+    unsigned remaining;
+    bool noCancellation;
+    while ((noCancellation = commData->lockFromCancellation()) && !(completed || error || (timedout = tm.timedout(&remaining))))
+    {
+        completed = commData->request.Test(stat);
+        commData->releaseCancellationLock();
+        usleep(WAIT_DELAY);
```

Comment:
this looks like a tight loop (albeit with a delay).
Will need to revisit this paradigm to look at blocking and be signalling instead.

Reply:
I saw in existing MP implementation something similar like what you said was done in several places. couldn't get a clear idea how it was done.  would be great if you could give me a simple example.